lib/auth/accountrecovery_test.go

```diff
@@ -111,3 +133,264 @@ func TestRecoveryCodeEventsEmitted(t *testing.T) {
+	})
+	require.NoError(t, err)
+	require.Equal(t, types.UserTokenUsage_RECOVER_PWD, startToken.GetUsage())
+
+	// Test with no recover type.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[2]),
+	})
+	require.Error(t, err)
+}
+
+func TestStartAccountRecovery_WithLock(t *testing.T) {
+	srv := newTestTLSServer(t)
+	ctx := context.Background()
+	fakeClock := srv.Clock().(clockwork.FakeClock)
+
+	defaultModules := modules.GetModules()
+	defer modules.SetModules(defaultModules)
+	modules.SetModules(&testWithCloudModules{})
+
+	u, err := createUserWithSecondFactorAndRecoveryCodes(srv, "otp")
+	require.NoError(t, err)
+
+	// Test max failed recovery attempt locks both login and further recovery attempt.
+	for i := 1; i <= defaults.MaxAccountRecoveryAttempts; i++ {
+		_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+			Username: u.username,
+		})
+		require.Error(t, err)
+	}
+
+	user, err := srv.Auth().GetUser(u.username, false)
+	require.NoError(t, err)
+	require.True(t, user.GetStatus().IsLocked)
+	require.False(t, user.GetStatus().LockExpires.IsZero())
+	require.False(t, user.GetStatus().RecoveryAttemptLockExpires.IsZero())
+
+	// Advance time to remove lock and attempts.
+	fakeClock.Advance(defaults.AttemptTTL)
+
+	// Trigger login lock.
+	for i := 1; i <= defaults.MaxLoginAttempts; i++ {
+		_, err = srv.Auth().authenticateUser(ctx, AuthenticateUserRequest{
+			Username: u.username,
+			OTP:      &OTPCreds{},
+		})
+		require.Error(t, err)
+
+		if i == defaults.MaxLoginAttempts {
+			require.True(t, trace.IsAccessDenied(err))
+		}
+	}
+
+	// Test recovery is still allowed after login lock.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[0]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.NoError(t, err)
+
+	// Trigger max failed recovery attempts.
+	for i := 1; i <= defaults.MaxAccountRecoveryAttempts; i++ {
+		_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+			Username: u.username,
+		})
+		require.Error(t, err)
+
+		// The third failed attempt should return error.
+		if i == defaults.MaxAccountRecoveryAttempts {
+			require.EqualValues(t, ErrMaxFailedRecoveryAttempts, err)
+		}
+	}
+
+	// Test recovery is denied from attempt recovery lock.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[1]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.True(t, trace.IsAccessDenied(err))
+}
+
+type userAuthCreds struct {
+	recoveryCodes []string
+	username      string
+	password      []byte
+	u2fKey        *mocku2f.Key
+}
+
+func createUserWithSecondFactorAndRecoveryCodes(srv *TestTLSServer, secondFactor string) (*userAuthCreds, error) {
```

Comment:
I have some code that looks like this in #8094 (see [configureForMFA](https://github.com/gravitational/teleport/blob/4853995fd96eeaafcfa0f9adbe5d5ea873e07610/lib/auth/auth_login_test.go#L324)). Remind me to unify this once all PRs go through, or feel free to leave a TODO for me here ("TODO(codingllama): ...").